src/service_discovery/mod.rs

```diff
@@ -105,35 +111,40 @@ impl ServiceDiscovery {
     }
 
     fn read(&mut self, core: &mut Core, poll: &Poll) {
-        let (bytes_rxd, peer_addr) = match self.socket.recv_from(&mut self.read_buf) {
-            Ok((bytes_rxd, peer_addr)) => (bytes_rxd, peer_addr),
-            // TODO(povilas): handle WouldBlock as well
-            Err(ref e) if e.kind() == ErrorKind::Interrupted => return,
-            Err(e) => {
-                debug!("ServiceDiscovery error in read: {:?}", e);
-                self.terminate(core, poll);
-                return;
-            }
-        };
-
-        let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
-            Ok(msg) => msg,
-            Err(e) => {
-                debug!("Bogus message serialisation error: {:?}", e);
-                return;
-            }
-        };
+        loop {
+            let (bytes_rxd, peer_addr) = match self.socket.recv_from(&mut self.read_buf) {
+                Ok((bytes_rxd, peer_addr)) => (bytes_rxd, peer_addr),
+                Err(ref e)
+                    if e.kind() == ErrorKind::Interrupted || e.kind() == ErrorKind::WouldBlock =>
+                {
+                    return
+                }
+                Err(e) => {
+                    debug!("ServiceDiscovery error in read: {:?}", e);
+                    self.terminate(core, poll);
+                    return;
+                }
+            };
 
-        match msg {
-            DiscoveryMsg::Request { guid } => {
-                if self.listen && self.guid != guid {
-                    self.reply_to.push_back(peer_addr);
-                    self.write(core, poll)
+            let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
+                Ok(msg) => msg,
+                Err(e) => {
+                    debug!("Bogus message serialisation error: {:?}", e);
+                    return;
```

Comment:
Will be a `continue` or NoOp (if there's nothing after the `recv_from()` match arm above, so it'll naturally continue the loop) once you have inlined it.